### PR TITLE
testing/llvm8: new aport

### PR DESCRIPTION
**Pull request by Cogitri**

### CODE REVIEW

**Comment by Cogitri** [comment]
I guess the patch in the ML will take preference over this.

**Comment by Cogitri** [comment]
https://patchwork.alpinelinux.org/patch/4844/

**Comment by Cogitri** [comment]
So if there's interest in getting this merged soon feel free to use this.

File: testing/llvm8/disable-FileSystemTest.CreateDir-perms-assert.patch [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+-  ASSERT_EQ(Status.permissions() & fs::perms::all_all,
+-            fs::perms::owner_read | fs::perms::owner_exe);
+-  ASSERT_NO_ERROR(fs::create_directory(Twine(TestDirectory) + "baz777", false,
+-                                       fs::perms::all_all));
+-  ASSERT_NO_ERROR(fs::status(Twine(TestDirectory) + "baz777", Status));
+-  ASSERT_EQ(Status.permissions() & fs::perms::all_all, fs::perms::all_all);
+-
+-  // Restore umask to be safe.
+-  ::umask(OldUmask);
++//  // Set a 0000 umask so that we can test our directory permissions.
```
Comment (tcely) [inline testing/llvm8/disable-FileSystemTest.CreateDir-perms-assert.patch]: This test is misunderstanding umask. Try 0777 instead.

File: testing/llvm8/APKBUILD [hunk 2]
```diff
@@ -0,0 +1,257 @@
+# Contributor: Martell Malone <martell@marinelayer.io>
+# Contributor: Travis Tilley <ttilley@gmail.com>
+# Contributor: Mitch Tishmack <mitch.tishmack@gmail.com>
+# Contributor: Jakub Jirutka <jakub@jirutka.cz>
+# Contributor: Rasmus Thomsen <oss@cogitri.dev>
+# Maintainer: Natanael Copa <ncopa@alpinelinux.org>
+_pkgname=llvm
+pkgver=8.0.0
+_majorver=${pkgver%%.*}
```
Comment (tcely) [inline testing/llvm8/APKBUILD]:
```suggestion
_majorver=${pkgver%${pkgver##[0-9]}}
```

File: testing/llvm8/APKBUILD [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+pkgname=$_pkgname$_majorver
+pkgrel=0
+pkgdesc="Low Level Virtual Machine compiler system, version $_majorver"
+arch="all"
+url="https://llvm.org/"
+license="NCSA"
+depends_dev="$pkgname=$pkgver-r$pkgrel"
+makedepends="binutils-dev chrpath cmake file libffi-dev paxmark python3 zlib-dev
```
Comment (tcely) [inline testing/llvm8/APKBUILD]: file should be installed already

File: testing/llvm8/APKBUILD [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+subpackages="$pkgname-static $pkgname-libs $pkgname-dev $pkgname-test-utils:_test_utils"
+source="https://llvm.org/releases/$pkgver/llvm-$pkgver.src.tar.xz
+	disable-FileSystemTest.CreateDir-perms-assert.patch
+	0001-Disable-dynamic-lib-tests-for-musl-s-dlclose-is-noop.patch
+	fix-memory-mf_exec-on-aarch64.patch
+	fix-LLVMConfig-cmake-install-prefix.patch
+	python3-test.patch
+	"
+builddir="$srcdir/$_pkgname-$pkgver.src"
+options="!checkroot"
```
Comment (tcely) [inline testing/llvm8/APKBUILD]: Why is `!checkroot` being used?

File: testing/llvm8/APKBUILD [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+# BTW, Fedora and Debian do the same thing.
+#
+_prefix="usr/lib/llvm$_majorver"
+
+prepare() {
+	default_prepare
+	mkdir -p "$builddir"/build
+
+	# Known broken test on musl
+	rm test/CodeGen/AArch64/wineh4.mir
```
Comment (tcely) [inline testing/llvm8/APKBUILD]:
```suggestion
	rm -v test/CodeGen/AArch64/wineh4.mir
```

File: testing/llvm8/APKBUILD [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+	if [ "$_default_llvm" = yes ]; then
+		replaces="llvm${subpkgname#$pkgname}"
+		provides="$replaces=$pkgver-r$pkgrel"
+	fi
+}
+
+_mv() {
+	local dest; for dest; do true; done  # get last argument
+	mkdir -p "$dest"
+	mv $@
```
Comment (tcely) [inline testing/llvm8/APKBUILD]:
```suggestion
	mv "$@"
```

File: testing/llvm8/APKBUILD [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+		fi
+		case "$name" in
+			FileCheck | obj2yaml | yaml2obj) continue;;
+		esac
+		ln -s ../lib/llvm$_majorver/bin/$name "$pkgdir"/usr/bin/$newname
+	done
+
+	# Move /usr/lib/$pkgname/include/ into /usr/include/$pkgname/
+	# and symlink it back.
+	_mv include/* "$pkgdir"/usr/include/$pkgname/
```
Comment (tcely) [inline testing/llvm8/APKBUILD]:
```suggestion
	_mv include "$pkgdir"/usr/include/$pkgname
```

File: testing/llvm8/APKBUILD [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+	rmdir include
```
Comment (tcely) [inline testing/llvm8/APKBUILD]:
```suggestion

```

File: testing/llvm8/APKBUILD [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+	ln -s ../../include/$pkgname include
+
+	# Move /usr/lib/$pkgname/lib/cmake/llvm/ into /usr/lib/cmake/$pkgname/
+	# and symlink it back.
+	_mv lib/cmake/llvm/* "$pkgdir"/usr/lib/cmake/$pkgname/
```
Comment (tcely) [inline testing/llvm8/APKBUILD]:
```suggestion
	mkdir -p "$pkgdir"/usr/lib/cmake
	mv lib/cmake/llvm "$pkgdir"/usr/lib/cmake/$pkgname
```

File: testing/llvm8/APKBUILD [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+	rmdir lib/cmake/llvm
```
Comment (tcely) [inline testing/llvm8/APKBUILD]:
```suggestion

```

**Comment by Cogitri** [comment]
Ah, sorry, I guess I should've taken a closer look at the LLVM7 APKBUILD from which this is mostly copied! Will fix it later this day. Thanks for the review!

**Comment by 5paceToast** [comment]
1. Tests expect bash and py3-setuptools.
    The latter appears to be installed incidentally in drone's environment, but that's not the general case (failed on my builder).
2. The test under question does not fail on my system.
    It would be interesting to see if it fails on the real builders (if it only fails within drone, it shouldn't be disabled, and the PR should get a ci-malfunction tag instead).

**Comment by Cogitri** [comment]
Pushed new version.

**Comment by Cogitri** [comment]
Gonna squash the commits once I've finished up setting up Alpine again on a ZFS encrypted root, is there something else I should change about this?

**Comment by Cogitri** [comment]
I can push new versions of clang, lldb and the other LLVM stuff once we have this moved to main.

**Pull request closed without merging**